Modify rule S4023: LayC and improve rule description (#2691)
SonarSource/rspec: rules/S4023/csharp/rule.adoc

== Why is this an issue?

Empty interfaces are either useless or used as a https://en.wikipedia.org/wiki/Marker_interface_pattern[marker]. https://learn.microsoft.com/en-us/dotnet/standard/attributes/writing-custom-attributes[Custom attributes] are a better alternative to marker interfaces.

A typical use case for marker interfaces is doing type inspection via https://learn.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/reflection[reflection] as shown below.

The `IIncludeFields` marker interface is used to configure the JSON serialization of an object.

[source,csharp,diff-id=1,diff-type=noncompliant]
----
// An example marker interface
public interface IIncludeFields { }

public class OptInToIncludeFields: IIncludeFields { }

Serialize(new OptInToIncludeFields());

void Serialize<T>(T o)
{
    // Use reflection to check if the interface is applied to the type
    var includeFields = o.GetType()
        .GetInterfaces().Any(i => i == typeof(IIncludeFields));
    var options = new JsonSerializerOptions()
    {
        // Take decisions based on the presents of the marker
        IncludeFields = includeFields, 
    };

> }

Reviewer: That's a good sample, just a thought: I would simplify it a bit and avoid all the JSON serialization configuration, I think it's not a relevant topic for this rule and might add complexity for the reader.

Author: If I remove that part, the `includeFields` variable is mysterious to the reader: What is this used for? By assigning it to `JsonSerializerOptions.IncludeFields` it is obvious to the reader and he can look it up to learn more about it. I think it is essential to the understanding of the `Serialize` method purpose.

        IncludeFields = includeFields,
    };
}
----

=== Exceptions

This rule doesn't raise in any of the following cases:

==== Aggregation of multiple interfaces

[source,csharp]
----
public interface IAggregate: IComparable, IFormattable { } // Compliant
----

==== Generic specialization

An empty interface with a single base interface is compliant as long as the resulting interface binds a generic parameter or constrains it.

[source,csharp]
----
// Compliant: bound to a concrete type
public interface IStringEquatable: IEquatable<string> { }
// Compliant: specialized by type parameter constraint
public interface ICreateableEquatable<T>: IEquatable<T> where T: new() { } 
----

==== Custom attribute

An empty interface with a single base interface is compliant if a custom attribute is applied to the interface.

[source,csharp]
----
[Obsolete]
public interface IConsoleLogger: ILogger { }
----

== How to fix it

Do any of the following to fix the issue:

* Add members to the interface
* Remove the useless interface
* Replace the usage as a marker interface with custom attributes

=== Code examples

==== Noncompliant code example

> [source,csharp,diff-id=2,diff-type=noncompliant]

Reviewer: It's a bit confusing to have one solution in the "Why section" and another solution in the "How to fix it".
I would keep both examples in this section, each under their appropriate bullet point.
- Add members to the interface:
[Noncompliant/Compliant snippet]
- Replace the usage as a marker interface with custom attributes
[Noncompliant/Compliant snippet]

> Interfaces describe functionality that an implementing class has to provide. An empty interface doesn't add any such functional requirements and is as such useless.

Reviewer: ```suggestion
Empty interfaces should be avoided as they do not provide any functional requirements for implementing classes.
```

== Why is this an issue?

Empty interfaces are either useless or used as a https://en.wikipedia.org/wiki/Marker_interface_pattern[marker]. https://learn.microsoft.com/en-us/dotnet/standard/attributes/writing-custom-attributes[Custom attributes] are a better alternative to marker interfaces. See the _How to fix it_ section for more information.

=== Exceptions

This rule doesn't raise in any of the following cases:

==== Aggregation of multiple interfaces

[source,csharp]
----

> public interface IAggregate: IComparable, IFormattable { } // Compliant

Reviewer: The first snippet has a `// Compliant` comment, the second an extended `// Compliant: explanation` and the third none. IMO we can improve a bit the consistency

=== Code examples

==== Noncompliant code example

The empty interface does not add any functionality.

[source,csharp,diff-id=1,diff-type=noncompliant]
----
public interface IFoo // Noncompliant
{ 

}
----

==== Compliant solution

Add members to the interface to be compliant.

[source,csharp,diff-id=1,diff-type=compliant]
----
public interface IFoo
{ 
    void Bar();
}
----

==== Noncompliant code example

A typical use case for marker interfaces is doing type inspection via https://learn.microsoft.com/en-us/dotnet/framework/reflection-and-codedom/reflection[reflection] as shown below.

The `IIncludeFields` marker interface is used to configure the JSON serialization of an object.

[source,csharp,diff-id=2,diff-type=noncompliant]
----
// An example marker interface
public interface IIncludeFields { }

public class OptInToIncludeFields: IIncludeFields { }

Serialize(new OptInToIncludeFields());

void Serialize<T>(T o)
{
    // Use reflection to check if the interface is applied to the type
    var includeFields = o.GetType()
        .GetInterfaces().Any(i => i == typeof(IIncludeFields));
    var options = new JsonSerializerOptions()
    {

>         // Take decisions based on the presents of the marker

Reviewer 2: ~presents~ presence

        IncludeFields = includeFields, 
    };
}
----

The same example can be rewritten using custom attributes. This approach is preferable because it is more fine-grained, allows parameterization, and is more flexible in type hierarchies.

[source,csharp,diff-id=2,diff-type=compliant]
----
// A custom attribute used as a marker
[AttributeUsage(AttributeTargets.Class)]
public sealed class IncludeFieldsAttribute: Attribute { }

[IncludeFields]
public class OptInToIncludeFields { }

Serialize(new OptInToIncludeFields());

void Serialize<T>(T o)
{
    // Use reflection to check if the attribute is applied to the type
    var includeFields = o.GetType()
        .GetCustomAttributes(typeof(IncludeFieldsAttribute), inherit: true).Any();
    var options = new JsonSerializerOptions()
    {

>         // Take decisions based on the presents of the marker

Reviewer 2: ~presents~ presence

Author: Thank you! #2755
